### DISCUSSION TITLE
how is theia cloud lsp client and server communicate?

### DISCUSSION THREAD

**Discussion by N4A502** [eclipse-theia/theia#9809] [Q&A]
as I understand, the gitpod server is useing theia, and running in a docker environment. but i wonder that， if the lsp client and server both running in the docker? how is it work with the brower? 
for example, how is the lsp server response value transfer to the brower?

**Comment by msujew** [comment]
Hi @hasegetc and thank you for your question,

communication inside of Theia is an interesting and complicated topic in itself. You're right in your assumption that both lsp client and server are running in the backend. Additionally, both the vscode extension that provides the language client, as well as the language server itself, run in a separate process independent of Theia. (The extension runs in the so-called _plugin host process_)

In order for vscode extensions to be able to communicate with Theia, the backend initializes a _json-rpc_ connection to the _plugin host process_ and replaces any `vscode` import calls with proxies that invoke the rpc connection. Afterwards, the backend Theia process is aware of any messages send by the vscode extension. This includes messages sent by the language server. Now the Theia frontend comes into play. As soon as you load Theia, a websocket connection is established between the frontend and the backend. All messages between the frontend and backend travel through this connection. The backend pushes the received lsp data to the frontend via the websocket connection. A channel based message pumping mechanism on the frontend makes sure that any received message ends up at the correct place. 

Then, all that's left to do is for monaco to display the data received by the frontend. Do you have any follow-up questions?

**Reply by N4A502** [reply]
@msujew thank you very much, the technology is wonderful. =could you point out some source code package that handle this thing? just the package is ok， I can find out the exactly place

**Reply by msujew** [reply]
Well, there's quite a lot of code handling the communication, but I'll try to link the most important parts:

* The [`WebSocketConnectionProvider`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/messaging/ws-connection-provider.ts) is handling the frontend side of things. Creating a new websocket connection on frontend startup and pumping message from and to [`WebSocketChannels`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/messaging/web-socket-channel.ts)
* On the backend, the [`MessagingContribution`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/node/messaging/messaging-contribution.ts) handles incoming websocket requests and creates its own web socket channels again.
* Method calls between frontend and backend, like fetching files and so on is usually done with the [`JsonRpcProxy`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/messaging/proxy-factory.ts). The [official Theia documentation](https://theia-ide.org/docs/json_rpc/) contains more information on that.

**Reply by N4A502** [reply]
thanks a lot

**Comment by MixusMinimax** [comment]
Hi! I am currently developing an lsp (vscode) extension. I really wanted both the client and server of that to run in the browser to make everything react very quickly. Should I use something like code-server instead of theia then, or is this json-rpc communication fast enough? In addition, I worry about what happens when multiple users use the editor, since all their individual workspaces spawn their own instance of all plugins, wouldn't the theia backend become a bottleneck?

**Reply by msujew** [reply]
Hey @MixusMinimax,

you can build a vscode web extension in theory, but it comes with a bunch of caveats (most importantly: no direct access to the file system). Given that latency in Theia is quite minimal (even less than in vscode, see https://github.com/eclipse-theia/theia/issues/10684), you usually have roundtrips of <50ms. 

> In addition, I worry about what happens when multiple users use the editor, since all their individual workspaces spawn their own instance of all plugins, wouldn't the theia backend become a bottleneck?

Theia is not a multi-tenant system (neither is vscode), and we don't offer any support for adopters who try to use it as a multi-tenant system. However, given that Theia works well in Electron (where a single user can spawn multiple workspaces/plugin hosts for a single backend), I don't assume that the backend will become a bottleneck there.

**Reply by MixusMinimax** [reply]
Thanks! I'll probably have to look into some kind of Kubernetes solution then.

**Reply by msujew** [reply]
There is kubernetes based solution available by one of the Theia contributors, take a look at [theia cloud](https://theia-cloud.io/).
